File: web/lib/slug.ts
/**
 * Builder-name → URL slug normalizer. Mirrors backend/app/util/slug.py
 * exactly so the link from a freshly-rendered report ("Prestige Estates
 * Ltd") resolves to the same `/builder/<slug>` page the backend serves.
 *
 * Keep this in lockstep with the Python implementation. If you change one
 * side, update the other — divergence breaks the link entirely.
 */
const SUFFIX_TOKENS = [
  "private-limited",
  "pvt-ltd",
  "pvt-limited",
  "limited",
  "ltd",
  "llp",
  "inc",
  "corporation",
  "corp",
  "company",
  "co",
  "group",
  "developers",
  "developer",
  "builders",
  "builder",
  "constructions",
  "construction",
  "infrastructure",
  "infra",
  "projects",
  "estates",
  "realty",
  "realtors",
  "properties",
  "homes",
];

export function toBuilderSlug(name: string | null | undefined): string | null {
  if (!name) return null;
  // Strip non-ASCII accents the same way Python's unicodedata.NFKD →
  // ascii('ignore') does for the common Latin-with-accents case.
  let s = name.normalize("NFKD").replace(/[̀-ͯ]/g, "");
  s = s.toLowerCase();
  s = s.replace(/[^a-z0-9]+/g, "-").replace(/^-+|-+$/g, "");
  if (!s) return null;
  let changed = true;
  while (changed) {
    changed = false;
    for (const suf of SUFFIX_TOKENS) {
      if (s.endsWith("-" + suf)) {
        s = s.slice(0, -(suf.length + 1));
        changed = true;
        break;
      }
      if (s === suf) return null;
    }
  }
  return s || null;
}

File: backend/app/util/slug.py
"""Deterministic slug helpers.

Builder profile pages (`/builder/<slug>`) need a stable mapping between a
free-text builder name as it appears on a portal ("Prestige Estates Projects
Ltd.", "PRESTIGE ESTATES", "Prestige  Estates  Projects") and a single URL
slug. We normalize aggressively so all three of those resolve to the same
public page and so the SEO surface doesn't fragment across casing or
punctuation noise.
"""
from __future__ import annotations

import re
import unicodedata

# Corporate / sectoral suffixes we strip after slugification. Indian
# developers are routinely listed with these in some portals and bare in
# others; dropping them collapses the obvious duplicates.
_SUFFIX_TOKENS = (
    "private-limited",
    "pvt-ltd",
    "pvt-limited",
    "limited",
    "ltd",
    "llp",
    "inc",
    "corporation",
    "corp",
    "company",
    "co",
    "group",
    "developers",
    "developer",
    "builders",
    "builder",
    "constructions",
    "construction",
    "infrastructure",
    "infra",
    "projects",
    "estates",
    "realty",
    "realtors",
    "properties",
    "homes",
)


def to_slug(name: str | None) -> str | None:
    """Normalize a free-text builder name to a stable URL slug.

    Returns None for empty / unusable input so callers can short-circuit
    without first-class None checks. Strips ASCII accents, lowercases,
    collapses any run of non-alphanumerics to a single hyphen, then
    repeatedly trims well-known corporate suffix tokens off the tail.
    """
    if not name:
        return None
    s = unicodedata.normalize("NFKD", name)
    s = s.encode("ascii", "ignore").decode("ascii")
    s = s.lower()
    s = re.sub(r"[^a-z0-9]+", "-", s).strip("-")
    if not s:
        return None
    # Repeatedly strip trailing suffix tokens. "foo-developers-pvt-ltd" →
    # "foo-developers-pvt" → "foo-developers" → "foo".
    changed = True
    while changed:
        changed = False
        for suf in _SUFFIX_TOKENS:
            if s.endswith("-" + suf):
                s = s[: -(len(suf) + 1)]
                changed = True
                break
            if s == suf:
                return None
    return s or None
